scylla/src/network/connection.rs

```diff
@@ -900,6 +902,30 @@ impl Connection {
         .await
     }
 
+    fn handle_result_metadata_new_id(
+        prepared_statement: &PreparedStatement,
+        query_response: &QueryResponse,
+    ) {
+        if let ResponseWithDeserializedMetadata::Result(ResultWithDeserializedMetadata::Rows(
+            rows_result,
+        )) = &query_response.response
+        {
+            if rows_result.0.metadata().id().is_some()
+                && rows_result.0.metadata().id()
+                    != prepared_statement.get_current_result_metadata().id()
+            {
+                // Metadata in response is either cached metadata extracted from statement, or a new extracted from response.
+                // If the id differs from the one in prepared statment, I see 2 possibilities:
+                // 1. Metadata changed, and was already updated on PreparedStatement by another execution.
+                // 2. Metadata changed, and was not yet updated on PreparedStatement.
+                // We could distinguish those two cases by checking id on cached_metadata, but it is not necessary,
+                // because in both cases we should, or at least may, update the metadata on PreparedStatement.
+                let metadata = rows_result.0.metadata().clone().into_owned();
+                prepared_statement.update_current_result_metadata(Arc::new(metadata));
```

Comment:
:question: Why should we update the metadata anyway?
Don't we prefer to elide another allocation here in case of a race with another execution?

Reply:
In the second mentioned case we should update metadata. In the first we don't have to, but we may (it just updates metadata to be the same).
Those code paths are not critical for performance, schema changes are one-off events. I don't want to complicate the code just to optimize non-hot-path, I prefer to keep ti simpler to understand.